### PR TITLE
#31: Provider name is now dependent of user role

### PR DESCRIPTION
**Pull request by oskarhinc**

### CODE REVIEW

File: src/main/java/org/benetech/servicenet/adapter/DataAdapterFactory.java [hunk 1]
```diff
@@ -18,15 +19,15 @@ public DataAdapterFactory(ApplicationContext context) {
     public Optional<SingleDataAdapter> getSingleDataAdapter(String providerName) {
         try {
             return Optional.of(context.getBean(providerName + DATA_ADAPTER_SUFFIX, SingleDataAdapter.class));
-        } catch (NoSuchBeanDefinitionException e) {
+        } catch (NoSuchBeanDefinitionException | BeanNotOfRequiredTypeException e) {
```
Comment (pgesek) [inline src/main/java/org/benetech/servicenet/adapter/DataAdapterFactory.java]: Is this new exception something we handle silently like this? Seems like a configuration issue we would need to signal on

File: src/main/java/org/benetech/servicenet/adapter/DataAdapterFactory.java [hunk 2]
```diff
@@ -18,15 +19,15 @@ public DataAdapterFactory(ApplicationContext context) {
...
             return Optional.empty();
         }
     }
 
     public Optional<MultipleDataAdapter> getMultipleDataAdapter(String providerName) {
         try {
             return Optional.of(context.getBean(providerName + DATA_ADAPTER_SUFFIX, MultipleDataAdapter.class));
-        } catch (NoSuchBeanDefinitionException e) {
+        } catch (NoSuchBeanDefinitionException | BeanNotOfRequiredTypeException e) {
```
Comment (pgesek) [inline src/main/java/org/benetech/servicenet/adapter/DataAdapterFactory.java]: same as above

File: src/main/java/org/benetech/servicenet/service/UserService.java [hunk 3]
```diff
@@ -318,6 +318,15 @@ public void removeNotActivatedUsers() {
         return authorityRepository.findAll().stream().map(Authority::getName).collect(Collectors.toList());
     }
 
+    /**
+     * @return true if any user is logged, and he has admin role
+     */
+    public boolean isAdmin() {
+        return getUserWithAuthorities().map(u -> u.getAuthorities().stream()
```
Comment (pgesek) [inline src/main/java/org/benetech/servicenet/service/UserService.java]: isAdmin can be a method of the user or UserDTO entity

File: src/main/java/org/benetech/servicenet/service/UserService.java [hunk 4]
```diff
@@ -318,6 +318,15 @@ public void removeNotActivatedUsers() {
         return authorityRepository.findAll().stream().map(Authority::getName).collect(Collectors.toList());
     }
 
+    /**
+     * @return true if any user is logged, and he has admin role
+     */
+    public boolean isAdmin() {
```
Comment (pgesek) [inline src/main/java/org/benetech/servicenet/service/UserService.java]: we should rename this method to: isCurrentUserAdmin

File: src/main/java/org/benetech/servicenet/service/impl/DocumentUploadServiceImpl.java [hunk 5]
```diff
@@ -114,4 +114,15 @@ public void delete(UUID id) {
...
         documentUploadRepository.deleteById(id);
     }
+
+    private String getRealProviderName(String currentProviderName) {
+        if (userService.isAdmin()) {
+            return currentProviderName;
+        }
+
+        //TODO: Replace with provider name connected to the user
+        return userService.getUserWithAuthorities().map(User::getLogin).orElseThrow(
```
Comment (pgesek) [inline src/main/java/org/benetech/servicenet/service/impl/DocumentUploadServiceImpl.java]: this piece of code is strange, but I guess it's a todo?

File: src/main/webapp/app/shared/auth/private-route.tsx [hunk 6]
```diff
@@ -7,6 +7,7 @@ import ErrorBoundary from 'app/shared/error/error-boundary';
 
 interface IOwnProps extends RouteProps {
   hasAnyAuthorities?: string[];
+  isAdmin: any;
```
Comment (pgesek) [inline src/main/webapp/app/shared/auth/private-route.tsx]: isn't this a boolean?

File: src/main/java/org/benetech/servicenet/adapter/DataAdapterFactory.java (hunk 1 above)
Comment (oskarhinc) [reply src/main/java/org/benetech/servicenet/adapter/DataAdapterFactory.java]: BeanNotOfRequiredTypeException will be thrown when we'll ask for SingleDataAdapter, but the adapter for this partner is a type of MultipleData. In this case, we're simply returning an empty optional. Should it be changed?

File: src/main/java/org/benetech/servicenet/service/impl/DocumentUploadServiceImpl.java (hunk 5 above)
Comment (oskarhinc) [reply src/main/java/org/benetech/servicenet/service/impl/DocumentUploadServiceImpl.java]: Yes, when User will be connected to the System Account, we will return its name here, but for now, it's just login

File: src/main/webapp/app/shared/auth/private-route.tsx (hunk 6 above)
Comment (oskarhinc) [reply src/main/webapp/app/shared/auth/private-route.tsx]: it is, thanks

File: src/main/java/org/benetech/servicenet/service/UserService.java (hunk 4 above)
Comment (oskarhinc) [reply src/main/java/org/benetech/servicenet/service/UserService.java]: ok

File: src/main/java/org/benetech/servicenet/service/UserService.java (hunk 3 above)
Comment (oskarhinc) [reply src/main/java/org/benetech/servicenet/service/UserService.java]: ok

File: src/main/java/org/benetech/servicenet/adapter/DataAdapterFactory.java (hunk 1 above)
Comment (pgesek) [reply src/main/java/org/benetech/servicenet/adapter/DataAdapterFactory.java]: If this is something that should not happen in normal usage (and I assume it shouldn't?) we should raise an exception

Comment (oskarhinc) [reply src/main/java/org/benetech/servicenet/adapter/DataAdapterFactory.java]: Ok, thanks

File: src/main/java/org/benetech/servicenet/service/impl/DocumentUploadServiceImpl.java (hunk 5 above)
Comment (oskarhinc) [reply src/main/java/org/benetech/servicenet/service/impl/DocumentUploadServiceImpl.java]: As it's already done, I've updated this PR to cover this.

File: src/main/java/org/benetech/servicenet/adapter/DataAdapterFactory.java [hunk 7]
```diff
@@ -17,18 +16,28 @@ public DataAdapterFactory(ApplicationContext context) {
     }
 
     public Optional<SingleDataAdapter> getSingleDataAdapter(String providerName) {
-        try {
-            return Optional.of(context.getBean(providerName + DATA_ADAPTER_SUFFIX, SingleDataAdapter.class));
-        } catch (NoSuchBeanDefinitionException | BeanNotOfRequiredTypeException e) {
+        AbstractDataAdapter adapter = getDataAdapter(providerName);
+        if (adapter instanceof SingleDataAdapter) {
+            return Optional.of((SingleDataAdapter) adapter);
```
Comment (pgesek) [inline src/main/java/org/benetech/servicenet/adapter/DataAdapterFactory.java]: You should use optional.ofNullable

Comment (oskarhinc) [reply src/main/java/org/benetech/servicenet/adapter/DataAdapterFactory.java]: I think it's not necessary, as it's already verified and can't be null (by instanceof).

Comment (pgesek) [reply src/main/java/org/benetech/servicenet/adapter/DataAdapterFactory.java]: Ok

**Pull request merged**